DAO/src/lib.rs

```diff
@@ -0,0 +1,698 @@
+            },
+            exec::gas_available() - GAS_RESERVE,
+            0,
+        );
+    }
+
+    // Aborts the membership proposal. It can be used in case when applicant is disagree with the requested shares or the details the proposer  indicated by the proposer
+    // Requirements:
+    // * `msg::source()` must be the applicant
+    // * The proposal must be membership proposal
+    // * The proposal can be aborted during only the abort window
+    // * The proposal must not be aborted yet
+    // Arguments:
+    // * `proposal_id`: the proposal ID
+    async fn abort(&mut self, proposal_id: u128) {
+        if !self.proposals.contains_key(&proposal_id) {
+            panic!("proposal does not exist");
+        }
+        if self.proposals.get(&proposal_id).unwrap().applicant != msg::source() {
+            panic!("caller must be applicant");
+        }
+        let proposal = self.proposals.get_mut(&proposal_id).unwrap();
+        if exec::block_timestamp() > proposal.starting_period + self.abort_window {
+            panic!("The abort window is over");
+        }
+        if proposal.aborted {
+            panic!("Proposal has already been aborted");
+        }
+
+        let amount = proposal.token_tribute;
+        proposal.token_tribute = 0;
+        proposal.aborted = true;
+
+        transfer_tokens(&self.approved_token_program_id, &msg::source(), amount).await;
+
+        msg::reply(
+            DaoEvent::Abort {
+                member: msg::source(),
+                proposal_id,
+                amount,
+            },
+            exec::gas_available() - GAS_RESERVE,
+            0,
+        );
+    }
+
+    // Assigns the admin position to new actor
+    // Requirements:
+    // * Only admin can assign new admin
+    // Arguments:
+    // * `new_admin`: valid actor ID
+    fn set_admin(&mut self, new_admin: &ActorId) {
+        if self.admin != msg::source() {
+            panic!("only admin can assign new admin");
+        }
+        if new_admin == &ZERO_ID {
+            panic!("new admin ID cant be zero");
+        }
+        self.admin = *new_admin;
+        msg::reply(
+            DaoEvent::AdminUpdated(*new_admin),
+            exec::gas_available() - GAS_RESERVE,
+            0,
+        );
+    }
+
+    // Sets the delegate key that is responsible for submitting proposals and voting
+    // The deleagate key defaults to member address unless updated
+    // Requirements:
+    // * `msg::source()` must be DAO member
+    // * The delegate key must not be zero address
+    // * A delegate key can be assigned only to one member
+    // Arguments:
+    // * `new_delegate_key`: the valid actor ID
+    fn update_delegate_key(&mut self, new_delegate_key: &ActorId) {
+        if !self.is_member(&msg::source()) {
+            panic!("account is not a DAO member");
+        }
+        if self.member_by_delegate_key.contains_key(new_delegate_key) {
+            panic!("cannot overwrite existing delegate keys");
```

Comment:
This is contradicting the name of function `update_delegate_key'. Is this only onetime function?

Reply:
No, for example,  I want to delegate my voting right to you, but another DAO member has already delegated you the right to vote. In that situation I can't use your address as delegate key.